Update create-parser.md (#2455)
splunk/splunk-connect-for-syslog: docs/gettingstarted/create-parser.md

# Create a parser
Whn you create parsers, SC4S can offload Splunk indexers by performing operations that would normally be performed during index time, including
linebreaking, source and sourcetype setting, and timestamping. Creating a parser also reduces the need for  corresponding add-ons on your indexers.

## Before you start
* Make sure you have read our [contribution standards](../CONTRIBUTING.md).
* For more background information on how filters and parsers work, read the [sources](../sources/index.md) documentation in this manual.
* Prepare your [environment](../developing/index.md).
* Create a new branch in the repository where you will apply your changes.

## Procure a raw log message

> If you already have a raw log message, you can skip this step. Otherwise, you need to extract one to have something to work with. You can do this in multiple ways, this section describes two methods. 

Reviewer: three methods:
1. tcpdump
2. wireshark
3. Raw log message in Splunk

	
```

### Procure a raw log message using Wireshark
You can also read the logs using Wireshark from the .pcap file. From Wireshark go to Statistics > Conversations, then click on 'Follow Stream':

![ws_conversation](../resources/images/ws_conv.png)

> ### Save your raw log message in Splunk or an archive

Reviewer: ```suggestion
### Procure a raw log message by saving it in Splunk
```

> Once you get your stream of messages, copy one of them. Note that in UDP there are not usually any message separators.

Reviewer: Please move it to `### Procure a raw log message using Wireshark`

See [Obtaining "On-the-wire" Raw Events](https://splunk.github.io/splunk-connect-for-syslog/main/troubleshooting/troubleshoot_resources/#obtaining-on-the-wire-raw-events).

## Create a unit test
To create a unit test, use the existing test case that is most similar to your use case. The naming convention is `test_vendor_product.py`.

> 1. Make sure that your log is being parsed correctly by creating a test case. 

Reviewer: Please see https://splunk.github.io/splunk-connect-for-syslog/2455/gettingstarted/create-parser/ , it needs newlines to render correctly

Assuming you have a raw message like this:
```bash 
<14>1 2022-03-30T11:17:11.900862-04:00 host - - - - Carbon Black App Control event:  text="File 'c:\program files\azure advanced threat protection sensor\2.175.15073.51407\winpcap\x86\packet.dll' [c4e671bf409076a6bf0897e8a11e6f1366d4b21bf742c5e5e116059c9b571363] would have blocked if the rule was not in Report Only mode." type="Policy Enforcement" subtype="Execution block (unapproved file)" hostname="CORP\USER" username="NT AUTHORITY\SYSTEM" date="3/30/2022 3:16:40 PM" ip_address="10.0.0.3" process="c:\program files\azure advanced threat protection sensor\2.175.15073.51407\microsoft.tri.sensor.updater.exe" file_path="c:\program files\azure advanced threat protection sensor\2.175.15073.51407\winpcap\x86\packet.dll" file_name="packet.dll" file_hash="c4e671bf409076a6bf0897e8a11e6f1366d4b21bf742c5e5e116059c9b571363" policy="High Enforcement - Domain Controllers" rule_name="Report read-only memory map operations on unapproved executables by .NET applications" process_key="00000433-0000-23d8-01d8-44491b26f203" server_version="8.5.4.3" file_trust="-2" file_threat="-2" process_trust="-2" process_threat="-2" prevalence="50"
```

> then do the following:

Reviewer: Can it be rather:

```
Assuming you have a raw message like this:

`<14>1 2022-03-30T11:17:11.900862-04:00 host - - - - Carbon Black App Control event:  text="File 'c:\program files\azure advanced threat protection sensor\2.175.15073.51407\winpcap\x86\packet.dll' [c4e671bf409076a6bf0897e8a11e6f1366d4b21bf742c5e5e116059c9b571363] would have blocked if the rule was not in Report Only mode." type="Policy Enforcement" subtype="Execution block (unapproved file)" hostname="CORP\USER" username="NT AUTHORITY\SYSTEM" date="3/30/2022 3:16:40 PM" ip_address="10.0.0.3" process="c:\program files\azure advanced threat protection sensor\2.175.15073.51407\microsoft.tri.sensor.updater.exe" file_path="c:\program files\azure advanced threat protection sensor\2.175.15073.51407\winpcap\x86\packet.dll" file_name="packet.dll" file_hash="c4e671bf409076a6bf0897e8a11e6f1366d4b21bf742c5e5e116059c9b571363" policy="High Enforcement - Domain Controllers" rule_name="Report read-only memory map operations on unapproved executables by .NET applications" process_key="00000433-0000-23d8-01d8-44491b26f203" server_version="8.5.4.3" file_trust="-2" file_threat="-2" process_trust="-2" process_threat="-2" prevalence="50"`

1. Make sure that the message is a valid python string, where escape characters are placed correctly.
2. Anonymize the data.
3. Rename functions.
4. Update index and sourcetype fields.
5. Extract and replace values with field names in the test string.

This example shows a test case for Vmware Carbonblack Protect device:

6. Now run the test:
...
7. The parsed log should appear in Splunk:
```

* Make sure that the message is a valid python string, where escape characters are placed correctly.
* Anonymize the data.
* Rename functions.
* Update index and sourcetype fields.
* Extract and replace values with field names in the test string.

This example shows a test case for Vmware Carbonblack Protect device:
```python
--8<---- "docs/resources/test_vmware_carbonblack_with_comments.py"
```
NOTE: It is a known issue that the test case will timeout when it starts. If it fails, just re-run it.

2. Now run the test:

`poetry run pytest test/test_vendor_product.py`

This test will spin up a Splunk instance on your localhost and forward the parsed message there.

3. The parsed log should appear in Splunk:
![parsed_log](../resources/images/parser_dev_splunk_first_run.png)

> In this example the message is being parsed as a generic *nix:syslog sourcetype and we have ensured that the fields in the messages are properly recognized. 

Reviewer: ```suggestion
In this example the message is being parsed as a generic `nix:syslog` sourcetype. This means that the message format complied with RFC standards, and SC4S could correctly identify the format fields in the message.
```

## Create a parser
To assign your messages to the proper index and sourcetype you will need to create a parser. Your parser must be declared in `package/etc/conf.d/conflib`. The naming convention is `app-type-vendor_product.conf`.

> 1. If you already have a similar parser, you can use it as a reference. In the parser, make sure you assign the proper sourcetype, index, vendor, product, and template. The template shows how your message should be parsed before sending them to Splunk.

Reviewer: Maybe 

```suggestion
1. If you find a similar parser in SC4S, you can use it as a reference. In the parser, make sure you assign the proper sourcetype, index, vendor, product, and template. The template shows how your message should be parsed before sending them to Splunk.
```
